modules/data.atmosphere/R/align_met.R

```diff
@@ -0,0 +1,298 @@
+# 
+# Returns a list called met.out with 2 levels that are matched in temporal resolution & number of ensembles
+# List Layers
+#  1. dat.train
+#  2. dat.source
+# Sublist Layers: time, met variables
+# -----------------------------------
+# Notes
+# -----------------------------------
+# 1. This assumes that both the trian and source data are in *at least* daily resolution
+# -----------------------------------
+
+#----------------------------------------------------------------------
+# Begin Function
+#----------------------------------------------------------------------
+align.met <- function(train.path, source.path, yrs.train=NULL, n.ens=NULL, pair.mems = TRUE, seed=Sys.Date(), verbose = FALSE) {
+  # Load required libraries
+  library(ncdf4)
+  library(lubridate)
+  
+  met.out <- list() # where the aligned data will be stored
+
+  # ---------------
+  # 1. Read in & organize training data
+  # ---------------
+  met.out[["dat.train"]] <- list()
+  # 1.a. Determine if we have an ensemble in the training path or if it's a single time series
+  if(length(dir(train.path, ".nc"))>0){ # we have a single time series
+    n.trn = 1 # Ignore how many input ensemble members we asked for, we only actually have 1 here
+    
+    files.train <- dir(train.path, ".nc")
+    
+    yrs.file <- strsplit(files.train, "[.]")
+    yrs.file <- matrix(unlist(yrs.file), ncol=length(yrs.file[[1]]), byrow=T)
+    yrs.file <- as.numeric(yrs.file[,ncol(yrs.file)-1]) # Assumes year is always last thing before the file extension
+    
+    if(!is.null(yrs.train)){
+      files.train <- files.train[which(yrs.file %in% yrs.train)]
+      yrs.file <- yrs.file[which(yrs.file %in% yrs.train)]
+    }
+    
+    # Loop through the .nc files putting everything into a list
+    print("Processing Training Data")
+    pb <- txtProgressBar(min=0, max=length(files.train), style=3)
+    for(i in 1:length(files.train)){
+      yr.now <- yrs.file[i]
+      
+      ncT <- nc_open(file.path(train.path, files.train[i]))
+      
+      # Set up the time data frame to help index
+      nday <- ifelse(leap_year(yr.now), 366, 365)
+      ntime <- length(ncT$dim$time$vals)
+      step.day <- nday/ntime
+      step.hr  <- step.day*24
+      stamps.hr <- seq(step.hr/2, by=step.hr, length.out=1/step.day) # Time stamps centered on period
+      
+      # Create a data frame with all the important time info
+      # center the hour step
+      df.time <- data.frame(Year=yr.now, DOY=rep(1:nday, each=1/step.day), Hour=rep(stamps.hr, ntime))
+      df.time$Date <- strptime(paste(df.time$Year, df.time$DOY, df.time$Hour, sep="-"), format=("%Y-%j-%H"), tz="UTC")
+      met.out$dat.train[["time"]] <- rbind(met.out$dat.train$time, df.time)
+      
+      # Extract the met info, making matrices with the appropriate number of ensemble members
+      for(v in names(ncT$var)){
+        df.tem <- matrix(rep(ncvar_get(ncT, v), n.trn), ncol=n.trn, byrow=F)
```

Comment:
ncvar_get can only get one variable at a time.  All the stuff with n.trn may not be necessary in this instance since I did some restructuring, but we do need to make sure it ends up as a matrix rather than a vector.